src/scripts/main.js

```diff
@@ -1,3 +1,318 @@
+  }
+}
+
+function isGameWon() {
+  if (finalCell) {
+    messageText('win');
+  }
+}
+
+function gameOver() {
+  if (!canTilesSlide()) {
+    messageText('lose');
+  }
+}
+
+function hasEmptyTile() {
+  for (let r = 0; r < rows; r++) {
+    for (let c = 0; c < columns; c++) {
+      if (board[r][c] === 0) {
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+function setTwoTiles() {
+  if (!hasEmptyTile()) {
+    return;
+  }
+
+  let found = false;
+
+  while (!found) {
+    const randomCellNumber = Math.random() <= 0.1 ? 4 : 2;
+    const r = Math.floor(Math.random() * rows);
+    const c = Math.floor(Math.random() * columns);
+
+    if (board[r][c] === 0) {
+      board[r][c] = randomCellNumber;
+
+      const tile = document.getElementById(`${r}-${c}`);
+
+      tile.innerText = `${randomCellNumber}`;
+      tile.classList.add(`field-cell--${randomCellNumber}`);
+      found = true;
+    }
+  }
+}
+
+startButton.addEventListener('click', () => {
+  for (let r = 0; r < rows; r++) {
+    for (let c = 0; c < columns; c++) {
+      const tile = document.getElementById(`${r}-${c}`);
+
+      tile.innerText = '';
+      tile.classList.value = 'field-cell';
+      board[r][c] = 0;
+    }
+
+    score = 0;
+    gameScore.innerText = 0;
+  }
+
+  setTwoTiles();
+  setTwoTiles();
+
+  message.classList.value = 'hidden';
+  startButton.classList.remove('start');
+  startButton.classList.add('restart');
+  startButton.innerText = 'Restart';
+});
+
+document.addEventListener('keyup', (keyEvent) => {
+  if (keyEvent.code === 'ArrowLeft') {
+    slideLeft();
+  } else if (keyEvent.code === 'ArrowRight') {
+    slideRight();
+  } else if (keyEvent.code === 'ArrowUp') {
+    slideUp();
+  } else if (keyEvent.code === 'ArrowDown') {
+    slideDown();
+  }
+
+  if (tilesMoved) {
+    setTwoTiles();
+    gameOver();
+    isGameWon();
+  }
+
+  tilesMoved = false;
+  gameScore.innerText = score;
+});
+
+function filterZero(row) {
+  return row.filter(currentNumber => currentNumber !== 0);
+}
+
+function slide(row) {
+  let currentRow = row;
+
+  currentRow = filterZero(currentRow);
+
+  for (let i = 0; i < currentRow.length - 1; i++) {
+    if (currentRow[i] === currentRow[i + 1]) {
+      currentRow[i] *= 2;
+      currentRow[i + 1] = 0;
+      score += currentRow[i];
+    }
+  }
+
+  currentRow = filterZero(currentRow);
+
+  while (currentRow.length < columns) {
+    currentRow.push(0);
+  }
+
+  return currentRow;
+}
+
+function canTilesSlide() {
+  if (hasEmptyTile()) {
+    return true;
+  }
+
+  for (let r = 0; r < rows; r++) {
+    for (let c = 0; c < columns; c++) {
+      const tile = board[r][c];
+
+      if (c > 0 && tile === board[r][c - 1]) {
+        return true;
+      }
+
+      if (c < columns - 1 && tile === board[r][c + 1]) {
+        return true;
+      }
+
+      if (r > 0 && tile === board[r - 1][c]) {
+        return true;
+      }
+
+      if (r < rows - 1 && tile === board[r + 1][c]) {
+        return true;
+      }
```

Comment:
all this conditions can combined in one:

```suggestion
      if (
        (c > 0 && tile === board[r][c - 1]) ||
        (c < columns - 1 && tile === board[r][c + 1])  ||
        (r > 0 && tile === board[r - 1][c]) ||
        (r < rows - 1 && tile === board[r + 1][c]) ||
        ) {
        return true;
      }
```